### ISSUE TITLE
Losses become nan after some number of epochs

### ISSUE THREAD

**Issue by Baukebrenninkmeijer** [DAI-Lab/SDGym#6]
I've been using your nicely implemented TableGAN and TGAN for some training, but after 10-15 my loss becomes Nan. 

```
12 step 1200
tensor(0.0802, device='cuda:0', grad_fn=<SubBackward0>)
tensor(6.9029, device='cuda:0', grad_fn=<NegBackward>) None

epoch
12 step 1250
tensor(0.1213, device='cuda:0', grad_fn=<SubBackward0>)
tensor(5.7503, device='cuda:0', grad_fn=<NegBackward>) None

epoch
12 step 1300
tensor(0.2494, device='cuda:0', grad_fn=<SubBackward0>)
tensor(3.1711, device='cuda:0', grad_fn=<NegBackward>) None
epoch
12 step 1350
tensor(0.1131, device='cuda:0', grad_fn=<SubBackward0>)
tensor(8.1661, device='cuda:0', grad_fn=<NegBackward>) None
epoch
12 step 1400
tensor(0.1474, device='cuda:0', grad_fn=<SubBackward0>)
tensor(6.2397, device='cuda:0', grad_fn=<NegBackward>) None
epoch 12
step
1450
tensor(nan, device='cuda:0', grad_fn=<SubBackward0>)
tensor(nan, device='cuda:0', grad_fn=<NegBackward>) None
epoch 12

step 1500

tensor(nan, device='cuda:0', grad_fn=<SubBackward0>)
tensor(nan, device='cuda:0', grad_fn=<NegBackward>) None
```

It seems to be possibly related with the data definition given as input, where some of the target domains were not covered by the data, for example when taking a subset. This seems to be inducing the effect earlier. But at the moment I am not using that and getting the Nans after about half an hour of training, which seems to be wierd. It's not visible in the text, but up to the Nan point, the convergence seems to be a normal GAN curve, first dipping and then steadily rising. 

I'm testing with your toy datasets now to see if the issue persists.

Is this something you have encountered in your testing? Any thoughts?

**Comment by Baukebrenninkmeijer** [comment]
Successfully reproduced on the adult dataset with your specified preprocessing. As verification, the changes in the last few commits in project/scaffolding have only been syntactic, right?

![image](https://user-images.githubusercontent.com/9077462/60419920-b9976180-9be6-11e9-9c6b-ac77e1f41cbf.png)

**Comment by Baukebrenninkmeijer** [comment]
Tried it on both the Census as well as the Adult dataset. Also tried using both the GMMTransformer and the BGMTransformer. Neither change seems to make a difference.

**Comment by Baukebrenninkmeijer** [comment]
Narrowed the problem down to something specifically with TGAN. Using the latest project_scaffolding the problem persists with both the census and adult dataset. TableGAN does not seem to have the problem. MedGAN pretraining takes a long time so I don't have any results for that yet.

![image](https://user-images.githubusercontent.com/9077462/60426162-61675c00-9bf4-11e9-843b-d15fa0a5d853.png)

**Comment by ManuelAlvarezC** [comment]
Hi @Baukebrenninkmeijer,

Thanks for reporting this, we will have a look and see how can it be fixed.

**Comment by Baukebrenninkmeijer** [comment]
Good to hear. I have the old TGAN working so I don't necessarily need this TGAN. Thus, I likely won't investigate any further into this issue either.

**Comment by kveerama** [comment]
@Baukebrenninkmeijer we just released the paper about SDGym [here](https://arxiv.org/pdf/1907.00503.pdf). This explains the latest TGAN model (the older version was here in this [repo](https://github.com/DAI-Lab/TGAN)) This paper creates a newer much better version (we should really call it CTGAN to differentiate it from TGAN- I can see how that can be confusing). Also, the paper explains the reasons behind SDGym and why we created it.

**Comment by Baukebrenninkmeijer** [comment]
Haha you guys beat me to it! I was hoping your paper would be finished after my thesis.

Could you maybe elaborate on the usage of the Gumbel softmax? I've been confused by when to and when not to use adapted softmax functions to overcome the differentiation problem. In the old TGAN paper you stated that it was not necessary to use the Gumbel softmax because the number of categories is fairly small, so the probability distribution can be generated directly with softmax. However, in the new TGAN paper you decide to use it to overcome the same stated problem.

**Comment by Baukebrenninkmeijer** [comment]
@kveerama
